anvil-polkadot: add impersonation

iulianbarbu opened the pull request:
## Motivation

Closes #242 .

## Solution

1. Copy the cheat manager from anvil under substrate_node/impersonation.rs.
2. Override secp256k1_ecdsa_recover versions of sp_io crypto host functions and use them for the wasm executor of the substrate runtime.
3. Override kecckak_256 sp_io's hashing host functions and use it for the wasm executor of the substrate runtime.
5. Integration tests that go through: random account impersonation, stopping account impersonation, auto account impersonation.

alindima commented on crates/anvil-polkadot/tests/it/utils.rs:
```diff
@@ -124,6 +130,27 @@ impl TestNode {
         StorageKey(key)
     }
 
+    /// Execute an ethereum transfer transaction.
+    pub async fn eth_transfer(
+        &mut self,
+        from: Address,
+        to: Address,
+        amount: U256,
+        block_nr: u32,
+    ) -> Result<H256, Box<dyn std::error::Error>> {
+        let tx_hash = unwrap_response::<H256>(
+            self.eth_rpc(EthRequest::EthSendTransaction(Box::new(WithOtherFields::new(
+                TransactionRequest::default().value(amount).from(from).to(to),
+            ))))
+            .await
+            .unwrap(),
+        )?;
+        self.wait_for_block_with_timeout(block_nr, std::time::Duration::from_secs(5))
+            .await
+            .unwrap();
+        Ok(tx_hash)
+    }
+
```
Alex also added a similar function, let's use that if possible and modify it if needed

alindima commented on crates/anvil-polkadot/src/substrate_node/host.rs:
```diff
@@ -0,0 +1,95 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Host function overrides for signature verification.
```
let's add a comment in this file explaining why both overrides for key recovery and hashing are needed and how they work together

iulianbarbu commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,104 @@
+    unwrap_response::<()>(node.eth_rpc(EthRequest::SetAutomine(true)).await.unwrap()).unwrap();
+    let alith_initial_balance = node.get_balance(alith_account.address(), latest_block).await;
+    let dest_initial_balance = node.get_balance(dest_h160, latest_block).await;
+    let tx_hash = node.eth_transfer(alith_addr, dest_addr, transfer_amount, 1).await.unwrap();
```
[e5ff766](https://github.com/paritytech/foundry-polkadot/pull/332/commits/e5ff766a41301a4319e4bbb5bd5f1c5005476156)

iulianbarbu commented on crates/anvil-polkadot/src/api_server/server.rs:
```diff
@@ -459,5 +510,6 @@ async fn create_revive_rpc_client(substrate_service: &Service) -> Result<EthRpcC
             panic!("Block subscription task failed: {err:?}",)
         }
     });
+
```
[bd5dae5](https://github.com/paritytech/foundry-polkadot/pull/332/commits/bd5dae5b2bd19871c31d3e1eaddd2008c5df0e3c)

iulianbarbu commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,104 @@
+            - AlloyU256::from(receipt_info.effective_gas_price * receipt_info.gas_used).inner(),
+        dest_balance
+    );
+}
```
[affa8e5](https://github.com/paritytech/foundry-polkadot/pull/332/commits/affa8e5b6b4a611823cd8b7b97dd43daba559777)

iulianbarbu commented:
> From @alindima offline:
> 
>     * try to reduce Cargo.lock size if possible
> 
>     * remove lock from cheats manager (it is accessed in RPC calls which are serialized)
> 
>     * rename cheats logic specifically to impersonation (we don't cheat other than impersonation)

> IIRC this could also be configured via CLI

cc @alindima 
- used `impersonation` naming instead of `cheats` everywhere (simplifying the manager to be dedicated to impersonation as well by removing the cheats state, which planned to be more generic and used for other cheats too), removed the inner lock and added support for auto impersonation setting via CLI: [61c25f0](https://github.com/paritytech/foundry-polkadot/pull/332/commits/61c25f0e9bddefb6d53c2dab1e9e8d0934dbe79e)
- reduced cargo.lock size: [31dcfd0](https://github.com/paritytech/foundry-polkadot/pull/332/commits/31dcfd05bb56abcafc06d510bcf8624ffbfe006e) + [4e60394](https://github.com/paritytech/foundry-polkadot/pull/332/commits/4e6039400cc2568e05d3b3475e845e9c352911cc)

@AlexandruCihodaru  given our discussion offline about the toml formatters, I added a .taplo.toml config file, same as the one used by taplo itself. I don't think it is a bad idea doing it at this point, since editors differ in how they treat identation so using same formatting will simplify reviews. Alternatively we can reuse same config as for polkadot-sdk, but I don't think it is important. I might be able to do it in a separate PR but I think it is a bit pointless to revert the changes (they aren't that many and also they are simple to review). Either way please lmk your thoughts.

AlexandruCihodaru commented on crates/anvil-polkadot/tests/it/utils.rs:
```diff
@@ -124,6 +130,28 @@ impl TestNode {
+        )?;
+
+        if let Some(block_nr) = block_number {
+            self.wait_for_block_with_timeout(block_nr, std::time::Duration::from_secs(5))
```
Is there a reason we wait so much for a new block? 
nit: Also maybe it would be worth to check that we are not in a no_mining state before waiting

AlexandruCihodaru commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+    // Assert on balances after first transfer.
+    let alith_final_balance = node.get_balance(alith_account.address(), latest_block).await;
+    let dest_final_balance = node.get_balance(dest_h160, latest_block).await;
+    let existential_deposit = U256::from_str_radix("100000000000000000000", 10).unwrap();
```
nit: make this a constant in utils.rs because it could be useful in other tests

AlexandruCihodaru commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+    let transaction =
+        TransactionRequest::default().value(transfer_amount).from(dest_addr).to(alith_addr);
+    let err = node.send_transaction(transaction.clone(), Some(3)).await.unwrap_err();
+    assert!(err.to_string().starts_with(
+        r#"Expected success but got error: RpcError { code: InternalError, message: "Account not found for address"#
+    ));
```
assert!(matches!

AlexandruCihodaru replied:
Or even better if you wish
```rust
pub fn unwrap_error(response: ResponseResult) -> Result<RpcError, Box<dyn std::error::Error>> {
    match response {
        ResponseResult::Error(err) => Ok(err),
        ResponseResult::Success(value) => {
            Err(format!("Expected error but got success: {value:?}").into())
        }
    }
}
```

to be used as:

```rust
let err = unwrap_error(response).unwrap();
assert_eq!(err.code, ErrorCode::InternalError);
assert_eq!(err.message, "Account not found for address");
```

AlexandruCihodaru commented on crates/anvil-polkadot/src/substrate_node/service.rs:
```diff
@@ -17,14 +17,22 @@ use polkadot_sdk::{
 
-pub type FullClient =
-    sc_service::TFullClient<Block, RuntimeApi, WasmExecutor<sp_io::SubstrateHostFunctions>>;
+use super::host::{PublicKeyToHashOverride, SenderAddressRecoveryOverride};
```
use `use crate::` for imports

AlexandruCihodaru commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+    let receipt_info = node.get_transaction_receipt(tx_hash).await;
+
+    // Assert on balances after first transfer.
+    let alith_final_balance = node.get_balance(alith_account.address(), latest_block).await;
```
nit: Could use None for BlockId as it is expanded when converting to the Revive struct anyway.

AlexandruCihodaru commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+    // Assert on balances after second transfer.
+    let alith_balance = node.get_balance(alith_account.address(), latest_block).await;
+    let dest_balance = node.get_balance(dest_h160, latest_block).await;
+    assert_eq!(alith_final_balance, alith_balance - transfer_amount);
```
nit: This is a little bit confusing, maybe take alith's balance before sending the transaction?

AlexandruCihodaru commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+        r#"Expected success but got error: RpcError { code: InternalError, message: "Account not found for address"#
+    ));
+
+    // Start impersonating any address now
+    unwrap_response::<()>(node.eth_rpc(EthRequest::AutoImpersonateAccount(true)).await.unwrap())
+        .unwrap();
+
+    // Transfer at block 3 (same as for previous failed transfer, which did not result in a block
+    // being produced).
+    let tx_hash = node.send_transaction(transaction, Some(3)).await.unwrap();
+    tokio::time::sleep(std::time::Duration::from_secs(2)).await;
+    let receipt_info = node.get_transaction_receipt(tx_hash).await;
+
+    // Assert on balances after third transfer.
+    let alith_balance = node.get_balance(alith_account.address(), latest_block).await;
+    let dest_balance = node.get_balance(dest_h160, latest_block).await;
+    assert_eq!(alith_final_balance, alith_balance - transfer_amount);
+    assert_eq!(
+        dest_final_balance
+            - transfer_amount
+            - AlloyU256::from(receipt_info.effective_gas_price * receipt_info.gas_used).inner(),
+        dest_balance
+    );
+
+    // Stop impersonating destination, and assert on error when retrying the same transfer.
+    unwrap_response::<()>(node.eth_rpc(EthRequest::AutoImpersonateAccount(false)).await.unwrap())
+        .unwrap();
+    let transfer_amount = U256::from_str_radix("10000000", 10).unwrap();
+    let transaction =
+        TransactionRequest::default().value(transfer_amount).from(dest_addr).to(alith_addr);
+    let err = node.send_transaction(transaction.clone(), Some(4)).await.unwrap_err();
+    assert!(err.to_string().starts_with(
+        r#"Expected success but got error: RpcError { code: InternalError, message: "Account not found for address"#
+    ));
+}
```
Create a new test for this.

AlexandruCihodaru reviewed (approved):
Amazing job! ❤️

alindima reviewed (commented):
Good job!

iulianbarbu commented on .config/.taplo.toml:
```diff
@@ -0,0 +1,62 @@
+  "target",
+]
+
+[formatting]
```
Revert here: [5d489e0](https://github.com/paritytech/foundry-polkadot/pull/332/commits/5d489e082d50fd94a753148817788c8c2a35b897)

iulianbarbu commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+    let alith_addr = Address::from(ReviveAddress::new(alith_account.address()));
+    let dest_addr = Address::random();
+    let dest_h160 = H160::from_slice(dest_addr.as_slice());
+    let transfer_amount = U256::from_str_radix("1600000000000000000", 10).unwrap();
```
Please let me know if you meant something differently: [bcab3f9](https://github.com/paritytech/foundry-polkadot/pull/332/commits/bcab3f9ef8ea8e2541e50aa2ea0c35b9dc403c75)

iulianbarbu commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,199 @@
+    // Create a random account with some balance.
+    unwrap_response::<()>(node.eth_rpc(EthRequest::SetAutomine(true)).await.unwrap()).unwrap();
+    let alith_initial_balance = node.get_balance(alith_account.address(), latest_block).await;
+    let dest_initial_balance = node.get_balance(dest_h160, latest_block).await;
```
Done here: [bcab3f9](https://github.com/paritytech/foundry-polkadot/pull/332/commits/bcab3f9ef8ea8e2541e50aa2ea0c35b9dc403c75)

iulianbarbu commented on crates/anvil-polkadot/tests/it/impersonation.rs:
```diff
@@ -0,0 +1,231 @@
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_auto_impersonate_via_rpc() {
```
they should be similar in practice, will look into it!

iulianbarbu commented:
Some unrelated tests are failing, will have to look into it too before merging.
